backend/msgpack/conn.go

```diff
@@ -99,23 +98,36 @@ func newConnection(addr string, opts ConnectionOptions) *connection {
 func (c *connection) Write(data []byte) error {
 	c.Lock()
 	if c.conn == nil {
-		if !c.checkReconnectWithLock() {
+		if err := c.checkReconnectWithLock(); err != nil {
 			c.numFailures++
 			c.Unlock()
-			return errNoActiveConnection
+			return err
 		}
 	}
+
 	writeErr := c.writeWithLockFn(data)
 	if writeErr == nil {
 		c.Unlock()
 		return nil
 	}
-	if err := c.connectWithLockFn(); err == nil {
+
+	var temp bool
+	if opErr, ok := writeErr.(*net.OpError); ok && opErr.Temporary() {
```

Comment:
Not sure why we'd immediately retry on non-temporary errors but fall back to the retry logic for temporary errors? Also, the errors in the `net` package do not set this consistently AFAICT so don't know how much value we'd get from that at the cost of increased implementation complexity.
 
I'd lean towards not further complicating the logic here and follow the same retry mechanism as before regardless of whether the error is temporary or not.

Reply:
Sounds good, admittedly I think I got a little carried away, definitely seems like a premature optimization.